File: python/VCFFillableRecord.py

# coding: utf-8
# VCFFillableRecord.py

from __future__ import annotations
from functools import reduce
from itertools import *
from collections import Counter

import VCF
from VCFFamily import *
from GenoRecord import GenoRecord
from VCFHeteroHomo import VCFHeteroHomo
from VCFImpFamilyRecord import FillType, VCFImpFamilyRecord
from TypeDeterminer import ParentComb
from common import *
from Genotype import Genotype


#################### VCFFillableRecord ####################

class VCFFillableRecord(VCFFamilyRecord):
	def __init__(self, pos: int, geno: list[int], index: int,
									type: FillType, pair: ParentComb,
									probs: list[VCF.Probs]) -> None:
		super().__init__(pos, geno)
		self.index: int = index
		self.type: FillType = type
		self.pair: ParentComb = pair
		self.probs: list[VCF.Probs] = probs
	
	def gt_from_parent(self, mat_from: int, pat_from: int) -> int:
		if mat_from == 0 or pat_from == 0:
			return Genotype.NA
		
		gt_from_mat = self.get_mat_allele(mat_from-1)
		gt_from_pat = self.get_pat_allele(pat_from-1)
		return gt_from_mat | (gt_from_pat << 1) | 4
	
	# patからはどのアレルが来るのか分からない
	def gt_from_mat(self, mat_from: int, i: int) -> int:
		prog_gt = self.unphased(i)
		# matからはこのアレルが来ているはず
		mat_allele = (self.mat_gt() >> (mat_from-1)) & 1
		pat_gt = self.pat_gt()
		if prog_gt == Genotype.UN_00:
			if mat_allele == 0:
				return Genotype.PH_00
			elif pat_gt < Genotype.PH_11:
				return Genotype.PH_10
			else:
				return Genotype.PH_11
		elif prog_gt == Genotype.UN_11:
			if mat_allele == 1:
				return Genotype.PH_11
			elif pat_gt == Genotype.PH_00:
				return Genotype.PH_00
			else:
				return Genotype.PH_01
		elif prog_gt == Genotype.UN_01:
			if mat_allele == 0:
				return Genotype.PH_01
			else:
				return Genotype.PH_10
		else:
			return Genotype.NA
	
	def gt_from_pat(self, pat_from: int, i: int) -> int:
		prog_gt = self.unphased(i)
		# matからはこのアレルが来ているはず
		pat_allele = (self.pat_gt() >> (pat_from-1)) & 1
		mat_gt = self.mat_gt()
		if prog_gt == Genotype.UN_00:
			if pat_allele == 0:
				return Genotype.PH_00
			elif mat_gt == Genotype.PH_11:
				return Genotype.PH_11
			else:
				return Genotype.PH_01
		elif prog_gt == Genotype.UN_11:
			if pat_allele == 1:
				return Genotype.PH_11
			elif mat_gt == Genotype.PH_00:
				return Genotype.PH_10
			else:
				return Genotype.PH_00
		elif prog_gt == Genotype.UN_01:
			if pat_allele == 0:
				return Genotype.PH_10
			else:
				return Genotype.PH_01
		else:
			return Genotype.NA
	
	def mat_from(self, i: int) -> int:
		if self.is_NA(i):
			return 0
		elif not self.is_mat_hetero():
			return 0
		else:
			a = self.get_mat_allele(0)
			# 文字列にしたときに、1文字目がどうなるか
			gt = self.geno[i]
			if gt == Genotype.UN_11 or gt == Genotype.PH_10 or gt == Genotype.PH_11:
				b = 1
			else:
				b = 0
			
			if a == b:
				return 1
			else:
				return 2
	
	def pat_from(self, i: int) -> int:
		if self.is_NA(i):
			return 0
		elif not self.is_pat_hetero():
			return 0
		else:
			# ここおかしい
			a = self.get_pat_allele(0)
			# 文字列にしたときに、3文字目がどうなるか
			gt = self.geno[i]
			if gt == Genotype.UN_00 or gt == Genotype.PH_00 or gt == Genotype.PH_10:
				b = 0
			else:
				b = 1
			
			if a == b:
				return 1
			else:
				return 2
	
	def inverse_prog_gt(self, gt: int, inv_mat: bool, inv_pat: bool) -> int:
		def inverse_allele(a: int, inv: bool) -> int:
			return 1 - a if inv else a
		
		allele1 = Genotype.get_allele(gt, 0)
		allele2 = Genotype.get_allele(gt, 1)
		return (4 | inverse_allele(allele1, inv_mat) |
					(inverse_allele(allele2, inv_pat) << 1))
	
	def inverse_prog_gts(self, prog_gts: list[int],
							inv_mat: bool, inv_pat: bool) -> list[int]:
		inv_prog_gts = [ self.inverse_prog_gt(gt, inv_mat, inv_pat)
												for gt in prog_gts ]
		return inv_prog_gts
	
	def inverse_parents_gts(self, inv_mat: bool, inv_pat: bool) -> None:
		# both must be hetero
		if inv_mat:
			self.geno[0] = Genotype.inverse(self.geno[0])
		if inv_pat:
			self.geno[1] = Genotype.inverse(self.geno[1])
	
	def is_near_prog_gts(self, gts: list[int]) -> bool:
		num = 0
		dist = 0
		for i in range(len(gts)):
			if self.geno[i+2] != Genotype.UN_01:
				num += 1
			# self.genoはnon-phasedの前提だがそれでよいか？
			if Genotype.unphased(gts[i]) != self.geno[i+2]:
				dist += 1
		return dist < max(1, num // 2)
	
	def modify_gts(self, new_prog_gts: list[int]) -> None:
		for inv_mat, inv_pat in product((False, True), repeat=2):
			inv_prog_gts = self.inverse_prog_gts(new_prog_gts, inv_mat, inv_pat)
			if self.is_near_prog_gts(inv_prog_gts):
				self.inverse_parents_gts(inv_mat, inv_pat)
				for i in range(2, len(self.geno)):
					self.geno[i] = inv_prog_gts[i-2]
				return
		for i in range(2, len(self.geno)):
			self.geno[i] = new_prog_gts[i-2]
	
	def modify_parents_type(self) -> None:
		mat_gt = self.mat_gt()
		pat_gt = self.pat_gt()
		if (self.pair != ParentComb.P00x11 and
				((mat_gt == Genotype.PH_00 and pat_gt == Genotype.PH_11) or
				 (mat_gt == Genotype.PH_11 and pat_gt == Genotype.PH_00))):
			self.pair = ParentComb.P00x11
	
	def from_which_chrom(self, i: int, mat: bool) -> int:
		j = 0 if mat else 1
		parent_allele = self.get_allele(j, 0)
		return 1 if parent_allele == self.get_allele(i, j) else 2
	
	@staticmethod
	def convert(records: list[VCFImpFamilyRecord],
								vcf: VCFHeteroHomo) -> list[VCFFillableRecord]:
		ref_vcf = vcf.vcf
		cols = ref_vcf.extract_columns(vcf.samples)
		new_records: list[VCFFillableRecord] = []
		for i, record in enumerate(records):
			type = record.get_fill_type()
			probs = ref_vcf.records[i].parse_PL(record.geno, cols)
			r = VCFFillableRecord(record.pos, record.geno,
										record.index, type, record.pair, probs)
			new_records.append(r)
		return new_records
	
	@staticmethod
	def merge(records: list[VCFFillableRecord],
								samples: list[str]) -> GenoRecord:
		pos = records[0].pos
		geno = records[0].geno[:]
		for record in records[1:]:
			geno.extend(record.geno)
		return GenoRecord(pos, geno)
	
	def decide_by_majority(self, GTs: list[int]) -> int:
		# 多数決できるなら多数決
		c = Counter(GTs)
		dic_num = Counter(c.values())
		max_num = max(dic_num.keys())
		if dic_num[max_num] == 1:	# 最多のGenotypeは一つ
			for GT, num in c.items():
				if num == max_num:
					return GT
			else:
				# ここには来ないはず
				return GT
		else:						# 複数ある場合は乱数的なものを使う
			candidate_GTs = []
			for GT, num in c.items():
				if num == max_num:
					candidate_GTs.append(GT)
			candidate_GTs.sort()
			# posをhash化してGTを選ぶ
			d = len(candidate_GTs)
			s = 0
			pos = self.pos
			while pos > 0:
				s += pos % 10
				pos //= 10
			i = s % d
			return candidate_GTs[i]
	
	# 親が00x11のときGTと違うなら0|0と1|1を入れ替える
	def swap_parents(self, pos: int, GT: int) -> None:
		if GT == self.geno[pos]:
			return
		elif GT in (Genotype.PH_00, Genotype.PH_11):
			is_mat_00 = (pos == 0) ^ (GT == Genotype.PH_11)
			self.geno[0] = Genotype.PH_00 if is_mat_00 else Genotype.PH_11
			self.geno[1] = Genotype.PH_11 if is_mat_00 else Genotype.PH_00
			# 子どもも入れ換えなければならない
			prog_GT = Genotype.PH_01 if is_mat_00 else Genotype.PH_10
			for i in range(2, len(self.geno)):
				self.geno[i] = prog_GT
	
	@staticmethod
	def decide_duplicated_Genotype(records: list[VCFFillableRecord],
											ps: list[tuple[int, int]]) -> int:
		# 全部./.ならそのまま
		# 子どもがいたらそれ
		# ./.を除いて全部同じだったらそれ
		# ./.と0/0 x 1/1で全部なら0/0 x 1/1で多数決
		# ./.と0/0 x 1/1を除いて全部同じだったらそれ
		# ./.と0/0 x 1/1を除いて複数種あれば多数決
		GTs = [ records[i].geno[j] for i, j in ps ]
		if all(GT == Genotype.NA for GT in GTs):
			return Genotype.NA
		
		# 子どもが優先
		for (i, j), GT in zip(ps, GTs):
			if j >= 2 and GT != Genotype.NA:
				return GT
		
		GTs2 = [ GT for GT in GTs if GT != Genotype.NA ]
		if is_all_same(GTs2):
			return GTs2[0]
		
		GTs3 = [ GT for (i, j), GT in zip(ps, GTs)
						if GT != Genotype.NA and
						   records[i].pair != ParentComb.P00x11 ]
		if not GTs3:	# 0/0 x 1/1しかない
			return records[0].decide_by_majority(GTs2)
		elif is_all_same(GTs3):
			return GTs3[0]
		else:
			return records[0].decide_by_majority(GTs3)
	
	@staticmethod
	def integrate_each_sample(records: list[VCFFillableRecord],
										ps: list[tuple[int, int]]) -> None:
		GT = VCFFillableRecord.decide_duplicated_Genotype(records, ps)
		
		for i, j in ps:
			# 0/0 x 1/1の親なら相手と入れ替えなければならない
			if records[i].pair == ParentComb.P00x11 and j <= 1:
				records[i].swap_parents(j, GT)
	
	@staticmethod
	def integrate(records: list[VCFFillableRecord], samples: list[str],
						pos_samples: list[list[tuple[int, int]]]) -> GenoRecord:
		pos = records[0].pos
		# 0|0 x 1|1ならひっくり返せる
		for ps in pos_samples:
			GTs = [ records[i].geno[j] for i, j in ps ]
			if not is_all_same(GTs):
				VCFFillableRecord.integrate_each_sample(records, ps)
		
		# ひっくり返した後にGTをまとめる
		geno: list[int] = []
		for ps in pos_samples:
			i, j = ps[0]
			geno.append(records[i].geno[j])
		return GenoRecord(pos, geno)

File: src/VCFFillableRecord.cpp
#include <sstream>
#include <cassert>

#include "../include/VCFFillableRecord.h"
#include "../include/VCFHeteroHomo.h"
#include "../include/common.h"

using namespace std;


//////////////////// VCFFillableRecord ////////////////////

VCFFillableRecord *VCFFillableRecord::copy() const {
	return new VCFFillableRecord(pos, geno, index, type, comb, probs);
}

int VCFFillableRecord::gt_from_parent(int mat_from, int pat_from) const {
	if(mat_from == 0 || pat_from == 0)
		return Genotype::NA;
	
	const int	gt_from_mat = get_mat_allele(mat_from-1);
	const int	gt_from_pat = get_pat_allele(pat_from-1);
	return gt_from_mat | (gt_from_pat << 1) | 4;
}

int VCFFillableRecord::gt_from_mat(int mat_from, size_t i) const {
	const int	prog_gt = unphased(i);
	// this allele must come from mat
	const int	mat_allele = get_allele(0, mat_from-1);
	const int	pat_gt = this->pat_gt();
	if(prog_gt == Genotype::UN_00) {
		if(mat_allele == 0)
			return Genotype::PH_00;
		else if(pat_gt < Genotype::PH_11)
			return Genotype::PH_10;
		else
			return Genotype::PH_11;
	}
	else if(prog_gt == Genotype::UN_11) {
		if(mat_allele == 1)
			return Genotype::PH_11;
		else if(pat_gt == Genotype::PH_00)
			return Genotype::PH_00;
		else
			return Genotype::PH_01;
	}
	else if(prog_gt == Genotype::UN_01) {
		if(mat_allele == 0)
			return Genotype::PH_01;
		else
			return Genotype::PH_10;
	}
	else {
		return Genotype::NA;
	}
}

int VCFFillableRecord::gt_from_pat(int pat_from, size_t i) const {
	const int	prog_gt = unphased(i);
	const int	pat_allele = get_allele(1, pat_from-1);
	const int	mat_gt = this->mat_gt();
	if(prog_gt == Genotype::UN_00) {
		if(pat_allele == 0)
			return Genotype::PH_00;
		else if(mat_gt == Genotype::PH_11)
			return Genotype::PH_11;
		else
			return Genotype::PH_01;
	}
	else if(prog_gt == Genotype::UN_11) {
		if(pat_allele == 1)
			return Genotype::PH_11;
		else if(mat_gt == Genotype::PH_00)
			return Genotype::PH_10;
		else
			return Genotype::PH_00;
	}
	else if(prog_gt == Genotype::UN_01) {
		if(pat_allele == 0)
			return Genotype::PH_10;
		else
			return Genotype::PH_01;
	}
	else {
		return Genotype::NA;
	}
}

int VCFFillableRecord::mat_from(size_t i) const {
	if(this->is_NA(i))
		return 0;
	else if(!this->is_mat_hetero())
		return 0;
	else {
		const int	a = this->get_mat_allele(0);
		// TODO: if the genotype is non-phased, something wrong.
		const int	gt = this->geno[i];
		const int	b = (gt == Genotype::UN_11 ||
						 gt == Genotype::PH_10 ||
						 gt == Genotype::PH_11) ? 1: 0;
		
		if(a == b)
			return 1;
		else
			return 2;
	}
}

int VCFFillableRecord::pat_from(size_t i) const {
	if(this->is_NA(i))
		return 0;
	else if(!this->is_pat_hetero())
		return 0;
	else {
		const int	a = this->get_pat_allele(0);
		// TODO: if the genotype is non-phased, something wrong.
		const int	gt = this->geno[i];
		const int	b = (gt == Genotype::UN_00 ||
						 gt == Genotype::PH_00 ||
						 gt == Genotype::PH_10) ? 0 : 1;
		
		if(a == b)
			return 1;
		else
			return 2;
	}
}

void VCFFillableRecord::inverse_parents_gts(bool inv_mat, bool inv_pat) {
	// both must be hetero
	if(inv_mat)
		this->geno[0] = Genotype::inverse(this->geno[0]);
	if(inv_pat)
		this->geno[1] = Genotype::inverse(this->geno[1]);
}

bool VCFFillableRecord::is_near_prog_gts(const vector<int>& gts) const {
	int	num = 0;
	int	dist = 0;
	for(size_t i = 0; i < gts.size(); ++i) {
		// TODO: modify
		if(geno[i+2] != Genotype::UN_01)
			num += 1;
		if(geno[i+2] == Genotype::UN_01) {
			if(!(gts[i] == Genotype::PH_01 || gts[i] == Genotype::PH_10))
				dist += 1;
		}
		else if(geno[i+2] == Genotype::UN_00) {
			if(gts[i] != Genotype::PH_00)
				dist += 1;
		}
		else if(geno[i+2] == Genotype::UN_11) {
			if(gts[i] != Genotype::PH_11)
				dist += 1;
		}
		else {
			dist += 1;
		}
		/*
		if(Genotype::is_NA(geno[i+2])
			dist += 1;
		else if(Genotype::unphased(gts[i]) != geno[i+2])
			dist += 1;
		*/
	}
	return dist < max(1, num / 2);
}

int VCFFillableRecord::distance(const vector<int>& gts1,
								const vector<int>& gts2) {
	int	dist = 0;
	for(size_t i = 0; i < gts1.size(); ++i) {
		if(!Genotype::is_same_gts(gts1[i], gts2[i]))
			dist += 1;
	}
	return dist;
}

void VCFFillableRecord::modify_gts(const vector<int>& new_prog_gts) {
	vector<int>	orig_gts(geno.begin() + 2, geno.end());
	int	min_dist = distance(new_prog_gts, orig_gts);
	vector<int>	min_gts = new_prog_gts;
	int	min_i = 0;
	for(int i = 1; i < 4; ++i) {
		const bool	inv_mat = (i & 2) == 2;
		const bool	inv_pat = (i & 1) == 1;
		const vector<int>	inv_prog_gts = inverse_prog_gts(new_prog_gts,
															inv_mat, inv_pat);
		const int	dist = distance(inv_prog_gts, orig_gts);
		if(dist < min_dist) {
			min_dist = dist;
			min_gts = inv_prog_gts;
			min_i = i;
		}
	}
	
	const bool	inv_mat = (min_i & 2) == 2;
	const bool	inv_pat = (min_i & 1) == 1;
	this->inverse_parents_gts(inv_mat, inv_pat);
	std::copy(min_gts.begin(), min_gts.end(), this->geno.begin() + 2);
}

int VCFFillableRecord::inverse_prog_gt(int gt,
										bool inv_mat, bool inv_pat) {
	if(Genotype::is_NA(gt))
		return Genotype::NA;
	
	if(gt == 0)
		gt = 4;
	else if(gt == 1)
		gt = 6;
	else if(gt == 2)
		gt = 7;
	
	const int	a1 = Genotype::get_allele(gt, 0);
	const int	a2 = Genotype::get_allele(gt, 1);
	return Genotype::from_alleles(Genotype::inverse_allele(a1, inv_mat),
								  Genotype::inverse_allele(a2, inv_pat));
}

vector<int> VCFFillableRecord::inverse_prog_gts(const vector<int>& prog_gts,
											bool inv_mat, bool inv_pat) const {
	vector<int>	inv_prog_gts;
	for(auto p = prog_gts.begin(); p != prog_gts.end(); ++p) {
		inv_prog_gts.push_back(inverse_prog_gt(*p, inv_mat, inv_pat));
	}
	return inv_prog_gts;
}

void VCFFillableRecord::modify_parents_type() {
	const int	mat_gt = this->mat_gt();
	const int	pat_gt = this->pat_gt();
	if(this->comb != ParentComb::P00x11 &&
			((mat_gt == Genotype::PH_00 && pat_gt == Genotype::PH_11) ||
			 (mat_gt == Genotype::PH_11 && pat_gt == Genotype::PH_00)))
		comb = ParentComb::P00x11;
}

int VCFFillableRecord::from_which_chrom(size_t i, bool is_mat) const {
	const int j = is_mat ? 0 : 1;
	const int	parent_allele = this->get_allele(j, 0);
	return parent_allele == this->get_allele(i, j) ? 1 : 2;
}

int VCFFillableRecord::hash(int d) const {
	int	hash_ = 0;
	int	n = this->get_pos();
	while(n > 0) {
		const int	r = n % 10;
		n /= 10;
		hash_ += r;
	}
	return hash_ % d;
}

int VCFFillableRecord::decide_by_majority(const vector<int>& GTs) const {
	// ex) [0/0, 0/1, 0/0, 1/1] -> { 0/0: 2, 0/1: 1, 1/1: 1 }
	map<int, int>	counter;
	for(auto p = GTs.begin(); p != GTs.end(); ++p)
		counter[*p] += 1;
	
	// ex) { 0/0: 2, 0/1: 1, 1/1: 1 } -> { 2: [0/0], 1: [0/1, 1/1] }
	map<int, vector<int>>	dic_num;
	for(auto p = counter.begin(); p != counter.end(); ++p)
		dic_num[p->second].push_back(p->first);
	
	pair<int, vector<int>>	max_pair;
	for(auto p = dic_num.begin(); p != dic_num.end(); ++p) {
		if(p->first > max_pair.first)
			max_pair = *p;
	}
	
	vector<int>	max_GTs = max_pair.second;
	std::sort(max_GTs.begin(), max_GTs.end());	// to match the Python version
	if(max_GTs.size() == 1)
		return max_GTs.front();
	else	// Use a random number if there is more than one candidates
		return max_GTs[this->hash((int)max_GTs.size())];
}

void VCFFillableRecord::swap_parents(size_t i, int GT) {
	if(GT == this->geno[i])
		return;
	
	if(GT == Genotype::PH_00 || GT == Genotype::PH_11) {
		const bool	is_mat_00 = (i == 0) ^ (GT == Genotype::PH_11);
		this->geno[0] = is_mat_00 ? Genotype::PH_00 : Genotype::PH_11;
		this->geno[1] = is_mat_00 ? Genotype::PH_11 : Genotype::PH_00;
		
		// swap genotyeps of progenies
		const int	prog_GT = is_mat_00 ? Genotype::PH_01 : Genotype::PH_10;
		for(size_t j = 2; j < this->geno.size(); ++j)
			this->geno[j] = prog_GT;
	}
}

// Determine a single Genotype
// when the Genotype of the same sample differs from family to family
int VCFFillableRecord::decide_duplicated_Genotype(
								const vector<VCFFillableRecord *>& records,
								const vector<pair<size_t, size_t>>& positions) {
	vector<int>	GTs;
	for(auto p = positions.begin(); p != positions.end(); ++p)
		GTs.push_back(records[p->first]->geno[p->second]);
	
	// if all genotype are ./., not change
	if(std::all_of(GTs.begin(), GTs.end(),
			static_cast<bool(*)(int)>(&Genotype::is_NA)))
		return Genotype::NA;
	
	// if the sample is a progeny of a family, use the genotype
	for(size_t i = 0; i < GTs.size(); ++i) {
		const auto	j = positions[i].second;	// jth(0-based) sample of family
		const int	GT = GTs[i];
		if(j >= 2 && !Genotype::is_NA(GT))		// j >= 2 means progeny
			return GT;
	}
	
	vector<int>	GTs_less_NA;
	for(auto p = GTs.begin(); p != GTs.end(); ++p) {
		if(!Genotype::is_NA(*p))
			GTs_less_NA.push_back(*p);
	}
	
	// If all Genotypes except ./. are the same Genotype, use that Genotype
	if(Common::is_all_same(GTs_less_NA))
		return GTs_less_NA.front();
	
	vector<int>	GTs_less_00x11;
	for(size_t i = 0; i < GTs.size(); ++i) {
		const auto	*record = records[positions[i].first];
		const int	GT = GTs[i];
		if(!Genotype::is_NA(GT) && record->comb != ParentComb::P00x11)
			GTs_less_00x11.push_back(GT);
	}
	
	if(GTs_less_00x11.empty())	// only 0/0 x 1/1
		return records.front()->decide_by_majority(GTs_less_NA);
	else if(Common::is_all_same(GTs_less_00x11))
		return GTs_less_00x11.front();
	else
		return records.front()->decide_by_majority(GTs_less_00x11);
}

vector<VCFFillableRecord *> VCFFillableRecord::convert(
									const vector<VCFImpFamilyRecord *>& records,
									const VCFHeteroHomo *vcf) {
	const VCFSmall	*ref_vcf = vcf->get_ref_vcf();
	const auto	cols = ref_vcf->extract_columns(vcf->get_samples());
	vector<VCFFillableRecord *>	new_records;
	for(size_t i = 0; i < records.size(); ++i) {
		const auto	*record = records[i];
		const auto	type = record->get_fill_type();
		const auto	*ref_record = ref_vcf->get_record(i);
		const auto	probs = ref_record->parse_PL(record->get_geno(), cols);
		auto	*r = new VCFFillableRecord(record->get_pos(), record->get_geno(),
											record->get_index(), type,
											record->get_comb(), probs);
		new_records.push_back(r);
	}
	return new_records;
}

GenoRecord *VCFFillableRecord::merge(const vector<VCFFillableRecord *>& records,
														const STRVEC& samples) {
	const ll	pos = records[0]->get_pos();
	vector<int>	new_geno;
	for(auto p = records.begin(); p != records.end(); ++p) {
		const vector<int>&	geno = (*p)->get_geno();
		new_geno.insert(new_geno.end(), geno.begin(), geno.end());
	}
	return new GenoRecord(pos, new_geno);
}

void VCFFillableRecord::integrate_each_sample(
								const vector<VCFFillableRecord *>& records,
								const vector<pair<size_t, size_t>>& positions) {
	const int	GT = decide_duplicated_Genotype(records, positions);
	for(auto p = positions.begin(); p != positions.end(); ++p) {
		VCFFillableRecord	*record = records[p->first];
		if(record->is_00x11() && p->second <= 1)
			record->swap_parents(p->second, GT);
	}
}

bool VCFFillableRecord::is_all_same_GT(
						const vector<VCFFillableRecord *>& records,
						const vector<pair<size_t, size_t>>& pos_samples) {
	if(pos_samples.size() == 1)
		return true;
	
	const auto	pos_sample = pos_samples.front();
	const int	GT0 = records[pos_sample.first]->geno[pos_sample.second];
	for(auto p = pos_samples.begin() + 1; p != pos_samples.end(); ++p) {
		const int	GT = records[p->first]->geno[p->second];
		if(GT != GT0)
			return false;
	}
	return true;
}

GenoRecord *VCFFillableRecord::integrate(
					const vector<VCFFillableRecord *>& records,
					const vector<string>& samples,
					const vector<vector<pair<size_t, size_t>>>& pos_samples) {
	// Exchangeable if one parent Genotype is 0|0 and the other is 1|1
	for(auto p = pos_samples.begin(); p != pos_samples.end(); ++p) {
		if(!is_all_same_GT(records, *p)) {
			integrate_each_sample(records, *p);
		}
	}
	
	const ll	pos = records[0]->get_pos();
	vector<int>	geno;
	for(auto p = pos_samples.begin(); p != pos_samples.end(); ++p) {
		const pair<int, int>&	pos = p->front();
		geno.push_back(records[pos.first]->geno[pos.second]);
	}
	return new GenoRecord(pos, geno);
}

// after imputed
int VCFFillableRecord::from_which_chrom(const VCFFillableRecord *record,
													size_t i, bool is_mat) {
	if(record == NULL)
		return 0;
	
	return record->from_which_chrom(i, is_mat);
}
